server/sonar-server/src/main/java/org/sonar/server/computation/component/PathAwareCrawler.java

```diff
@@ -28,75 +29,96 @@
  * parent's.
  * As for {@link DepthTraversalTypeAwareCrawler}, this crawler supports max depth visit and ordering.
  */
-public abstract class PathAwareCrawler<T> extends PathAwareVisitorAdapter<T> implements ComponentCrawler {
+public final class PathAwareCrawler<T> implements ComponentCrawler {
 
+  private final PathAwareVisitor<T> visitor;
   private final DequeBasedPath<T> stack = new DequeBasedPath<>();
 
-  public PathAwareCrawler(Component.Type maxDepth, ComponentVisitor.Order order, StackElementFactory<T> factory) {
-    super(maxDepth, order, factory);
+  public PathAwareCrawler(PathAwareVisitor<T> visitor) {
+    this.visitor = requireNonNull(visitor);
   }
 
   @Override
   public void visit(Component component) {
-    if (component.getType().isDeeperThan(getMaxDepth())) {
+    if (!verifyDepth(component)) {
       return;
     }
 
     stack.add(new PathElementImpl<>(component, createForComponent(component)));
 
-    if (getOrder() == PRE_ORDER) {
+    if (this.visitor.getOrder() == PRE_ORDER) {
       visitNode(component);
     }
 
     visitChildren(component);
 
-    if (getOrder() == POST_ORDER) {
+    if (this.visitor.getOrder() == POST_ORDER) {
       visitNode(component);
     }
 
     stack.pop();
   }
 
+  private boolean verifyDepth(Component component) {
+    CrawlerDepthLimit maxDepth = this.visitor.getMaxDepth();
+    return maxDepth.isSameAs(component.getType()) || maxDepth.isDeeperThan(component.getType());
+  }
+
   private void visitChildren(Component component) {
     for (Component child : component.getChildren()) {
-      if (!child.getType().isDeeperThan(getMaxDepth())) {
+      if (verifyDepth(component)) {
         visit(child);
       }
     }
   }
 
   private void visitNode(Component component) {
-    visitAny(component, stack);
+    this.visitor.visitAny(component, stack);
     switch (component.getType()) {
       case PROJECT:
-        visitProject(component, stack);
+        this.visitor.visitProject(component, stack);
         break;
       case MODULE:
-        visitModule(component, stack);
+        this.visitor.visitModule(component, stack);
         break;
       case DIRECTORY:
-        visitDirectory(component, stack);
+        this.visitor.visitDirectory(component, stack);
         break;
       case FILE:
-        visitFile(component, stack);
+        this.visitor.visitFile(component, stack);
+        break;
+      case VIEW:
+        this.visitor.visitView(component, stack);
+        break;
+      case SUBVIEW:
+        this.visitor.visitSubView(component, stack);
+        break;
+      case PROJECT_VIEW:
+        this.visitor.visitProjectView(component, stack);
         break;
       default:
-        visitUnknown(component, stack);
+        this.visitor.visitUnknown(component, stack);
     }
   }
 
   private T createForComponent(Component component) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "createForComponent" is 16 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 8, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)